### PR TITLE
[Fix] Infinite while loop when called on root folder, add tests

### PR DESCRIPTION
**Pull request by BrettCleary**
When calling `getFirstExistingParentPath` on the parent folder of a non-existent root drive on Windows, the while loop proceeds indefinitely, crashing the calling program

This is fixed by early return if one iteration of the loop does not change the parent folder

This also changes the while loop to a for loop proceeding up until the max # of folders (with the same condition to stop as the while loop). This is much less error prone

Tests have been added and two other tests have been modified to run on windows as well

### CODE REVIEW

File: test/__helpers__/mockDependencies.ts [hunk 1]
```diff
@@ -12,7 +12,7 @@ function mockDependencies(overrides?: Partial<Dependencies>, options?: {
 		release: '11.5.0',
 		fsAccess: () => Promise.resolve(),
 		pathNormalize: normalize,
-		pathSep: '/',
+		pathSep: sep,
```
Comment (Alex-D) [inline test/__helpers__/mockDependencies.ts]: This means tests work differently depending on the platform we are running on.
So, I prefer to stick to the `'/'` value and switch to the `'\'` value if we test things for Windows.

File: src/functions/getFirstExistingParentPath.ts [hunk 2]
```diff
@@ -8,15 +8,31 @@ import Dependencies from '@/src/types/dependencies'
...
 
-	while (!parentDirectoryFound) {
-		parentDirectoryPath = dependencies.pathNormalize(parentDirectoryPath + '/..')
+	const FAILED_TO_FIND_EXISTING_DIRECTORY_VALUE = ''
+
+	/**
+	 * Linux max file path length is 4096 characters.
+	 * With / separators and 1 letter folder names, this gives us a max of ~2048 folders to traverse.
+	 * This is much less error prone than a while loop.
+	 */
```
Comment (Alex-D) [inline src/functions/getFirstExistingParentPath.ts:20]: This limit is not valid on every file system
https://en.wikipedia.org/wiki/Comparison_of_file_systems#Limits

So... I do not know if I want to add that to this loop.

File: test/functions/getFirstExistingParentPath.spec.ts [hunk 3]
```diff
@@ -1,24 +1,44 @@
...
+const getDependencies = (parentPath: string) => mockDependencies({
+	fsAccess: async (directoryPath: PathLike) => directoryPath === parentPath ? Promise.resolve() : Promise.reject(new Error('File does not exists')),
+})
 
 test('unix: get first existing parent path', async t => {
-	const parentPath = '/home/Alex'
-	const dependencies = mockDependencies({
-		fsAccess: async (directoryPath: PathLike) => directoryPath === parentPath ? Promise.resolve() : Promise.reject(new Error('File does not exists')),
-	})
+	const parentPath = normalize('/home/Alex')
```
Comment (Alex-D) [inline test/functions/getFirstExistingParentPath.spec.ts]: Why normalize the file path for a UNIX test since it's already a Unix-style path?

File: test/functions/getFirstExistingParentPath.spec.ts [hunk 4]
```diff
@@ -1,24 +1,44 @@
...
+
 import test from 'ava'
 import { PathLike } from 'fs'
 
 import getFirstExistingParentPath from '@/src/functions/getFirstExistingParentPath'
 import mockDependencies from '@/test/__helpers__/mockDependencies'
 
+const getDependencies = (parentPath: string) => mockDependencies({
+	fsAccess: async (directoryPath: PathLike) => directoryPath === parentPath ? Promise.resolve() : Promise.reject(new Error('File does not exists')),
+})
```
Comment (Alex-D) [inline test/functions/getFirstExistingParentPath.spec.ts:9]: Nice refactor :)
👍🏻

File: test/functions/getFirstExistingParentPath.spec.ts [hunk 5]
```diff
@@ -1,24 +1,44 @@
+import { normalize } from 'node:path'
+
 import test from 'ava'
 import { PathLike } from 'fs'
 
 import getFirstExistingParentPath from '@/src/functions/getFirstExistingParentPath'
 import mockDependencies from '@/test/__helpers__/mockDependencies'
 
+const getDependencies = (parentPath: string) => mockDependencies({
```
Comment (Alex-D) [inline test/functions/getFirstExistingParentPath.spec.ts:7]: Maybe consider renaming that with the mocking notion in mind:

```suggestion
const getMockedDependencies = (parentPath: string) => mockDependencies({
```

File: test/functions/getFirstExistingParentPath.spec.ts [hunk 6]
```diff
@@ -1,24 +1,44 @@
...
-	const dependencies = mockDependencies({
-		fsAccess: async (directoryPath: PathLike) => directoryPath === parentPath ? Promise.resolve() : Promise.reject(new Error('File does not exists')),
-	})
+	const parentPath = normalize('/home/Alex')
+	const dependencies = getDependencies(parentPath)
 
 	t.is(await getFirstExistingParentPath(parentPath, dependencies), parentPath)
 })
+
+test('win32: Gets parent to C:\\HyperPlay', async t => {
```
Comment (Alex-D) [inline test/functions/getFirstExistingParentPath.spec.ts]: Please remove reference to any existing platform

File: test/functions/getFirstExistingParentPath.spec.ts [hunk 7]
```diff
@@ -1,24 +1,44 @@
...
+	const parentPath = normalize('C:\\')
+	const dependencies = getDependencies(parentPath)
+	t.is(await getFirstExistingParentPath(normalize('C:\\HyperPlay'), dependencies), parentPath)
```
Comment (Alex-D) [inline test/functions/getFirstExistingParentPath.spec.ts]: Please remove reference to any existing platform

File: test/functions/getFirstExistingParentPath.spec.ts [hunk 8]
```diff
@@ -1,24 +1,44 @@
...
-	})
+	const parentPath = normalize('/home/Alex')
+	const dependencies = getDependencies(parentPath)
 
 	t.is(await getFirstExistingParentPath(parentPath, dependencies), parentPath)
 })
+
+test('win32: Gets parent to C:\\HyperPlay', async t => {
+	const parentPath = normalize('C:\\')
+	const dependencies = getDependencies(parentPath)
```
Comment (Alex-D) [inline test/functions/getFirstExistingParentPath.spec.ts:35]: I suggest to define the `pathSep` somewhere here.

File: test/functions/getFirstExistingParentPath.spec.ts [hunk 9]
```diff
@@ -1,24 +1,44 @@
...
-		fsAccess: async (directoryPath: PathLike) => directoryPath === parentPath ? Promise.resolve() : Promise.reject(new Error('File does not exists')),
-	})
+	const parentPath = normalize('/home/Alex')
+	const dependencies = getDependencies(parentPath)
 
 	t.is(await getFirstExistingParentPath(parentPath, dependencies), parentPath)
 })
+
+test('win32: Gets parent to C:\\HyperPlay', async t => {
+	const parentPath = normalize('C:\\')
```
Comment (Alex-D) [inline test/functions/getFirstExistingParentPath.spec.ts]: Do not normalize here neither

**Review by Alex-D — CHANGES_REQUESTED** [review]

File: test/__helpers__/mockDependencies.ts (hunk 1 above)
Comment (BrettCleary) [reply test/__helpers__/mockDependencies.ts]: ~~The paths were already being normalized in `getFirstExistingParentPath`, so the tests were already running differently depending on the platform. Normalizing on windows changes `'/'` to `'\'` with `dependencies.pathNormalize`.~~

~~This means tests can't be run on Windows runners, which was one of the goals of this PR, since the bug was crashing both [HyperPlay](https://github.com/HyperPlay-Gaming/hyperplay-desktop-client) and [Heroic](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher) when running on Windows.~~

File: src/functions/getFirstExistingParentPath.ts (hunk 2 above)
Comment (BrettCleary) [reply src/functions/getFirstExistingParentPath.ts:20]: Is there a number we can change it to that would accomplish this goal? Or another way to avoid a while loop/provide a failsafe if we miss another edge case?

File: test/functions/getFirstExistingParentPath.spec.ts (hunk 3 above)
Comment (BrettCleary) [reply test/functions/getFirstExistingParentPath.spec.ts]: ~~True. Do you think we should check the platform and only enable these tests when `os.platform() !== 'win32'`?~~

Comment (BrettCleary) [reply test/functions/getFirstExistingParentPath.spec.ts]: I see how you are doing win32 tests in the index.spec.ts now and will follow that

**Comment by BrettCleary** [comment]
@Alex-D apologies for the windows specific runner tests. it was wrong of me to assume that was a goal of this project.

The main issue is the infinite while loop when called on a root directory, which is platform agnostic and this PR fixes and the new tests confirm this.

I've updated the logic and tests to match what you were doing previously with win32 tests in `index.spec.ts`

File: test/__helpers__/mockDependencies.ts (hunk 1 above)
Comment (BrettCleary) [reply test/__helpers__/mockDependencies.ts]: reverted :ok_hand:

File: src/functions/getFirstExistingParentPath.ts (hunk 2 above)
Comment (BrettCleary) [reply src/functions/getFirstExistingParentPath.ts:20]: Ultimately this is of course your call and I'm happy to revert to the while loop if you would like

File: test/functions/getFirstExistingParentPath.spec.ts (hunk 8 above)
Comment (BrettCleary) [reply test/functions/getFirstExistingParentPath.spec.ts:35]: I am not following. Is this still relevant after the changes I committed?
